### DISCUSSION TITLE
error in getting public url and its showing undefined

### DISCUSSION THREAD

**Discussion by Rohit11222** [supabase/supabase#22346] [Questions]
I have a file upload.html where i am uploading my video and trying to get the URL to play it on the gnere.html page where i have a video player. the video is getting uploaded on the storage but the public URL is showing undefined on the upload.html page and the codes i provided are without keys

this is my upload.html page code

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Video Upload Page</title>
  <link rel="stylesheet" href="../css/styles.css">
</head>
<body>
  <header style="background-color: white;">
    <h1 style="color: black;">Video Upload Page</h1>
  </header>
  <main>
    <section>
      <form id="uploadForm" enctype="multipart/form-data">
        <input type="file" name="videoFile" id="videoFile" accept="video/*">
        <button type="submit" id="uploadButton">Upload Video</button>
      </form>
    </section>
  </main>
  <footer id="samosa">
    <p>&copy; 2024 Your Website. All rights reserved.</p>
  </footer>
  <script type="module">
    import { createClient } from 'https://cdn.jsdelivr.net/npm/@supabase/supabase-js@2/+esm'

    // Supabase initialization
    const supabaseUrl = '';
    const supabaseKey = '';
    const supabase = createClient(supabaseUrl, supabaseKey);

    // Service key or service role API key
    const serviceKey = ;

    // Set the authentication session with the service key
    const session = supabase.auth.setSession({ access_token: serviceKey });

    const uploadForm = document.getElementById('uploadForm');
    uploadForm.addEventListener('submit', async (event) => {
      event.preventDefault();
      const videoFile = document.getElementById('videoFile').files[0];
      if (videoFile) {
        try {
          console.log('Uploading video:', videoFile.name);
          // Upload the video file to the "uploads/" directory
          const { data, error } = await supabase.storage
            .from('videos')
            .upload(`uploads/${videoFile.name}`, videoFile, {
              upsert: true, // Allow overwriting existing files
            });

          if (error) {
            console.error('Error uploading video:', error);
          } else {
            console.log('Video uploaded successfully:', data);
            // Get the public URL of the uploaded video
            const { publicURL } = supabase.storage.from('uploads').getPublicUrl(data.path);
            console.log('Video public URL:', publicURL);
            // Redirect to genre.html with the video URL as a query parameter
            window.location.href = `genre.html?videoUrl=${encodeURIComponent(publicURL)}`;
          }
        } catch (error) {
          console.error('Error during upload:', error);
        }
      }
    });
  </script>
</body>
</html> 

this is my genre.html page code

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Browse by Genre</title>
  <link rel="stylesheet" href="../css/styles.css">
  <link rel="stylesheet" href="https://cdn.plyr.io/3.7.3/plyr.css">
</head>
<body>
  <header class="site-header">
    <h1 style="color:black">Unleash Magic</h1>
  </header>
  <main>
    <section id="videoPlayerSection">
      <video id="player" controls></video>
    </section>
    <div class="genre-dropdown-container">
      <select class="genre-dropdown" id="dropdown">
        <option value="pop">pop</option>
        <option value="hiphop">hiphop</option>
        <option value="cinematic">cinematic</option>
        <option value="rock">rock</option>
      </select>
    </div>
    <section id="songList">
      <h2>Songs</h2>
      <div id="songContainer"></div>
    </section>
  </main>
  <footer id="samosa">
    <p class="copyright">&copy; 2024 Your Website. All rights reserved.</p>
  </footer>
  <script src="https://cdn.plyr.io/3.7.3/plyr.js"></script>
  <script type="module">
    import { createClient } from 'https://cdn.jsdelivr.net/npm/@supabase/supabase-js@2/+esm'

    // Supabase initialization
    const supabaseUrl = '';
    const supabaseKey = '';
    const supabase = createClient(supabaseUrl, supabaseKey);

    // Service key or service role API key
    const serviceKey = '';

    // Set the authentication session with the service key
    const session = supabase.auth.setSession({ access_token: serviceKey });

    const dropdown = document.getElementById('dropdown');
    const songContainer = document.getElementById('songContainer');
    const videoPlayer = document.getElementById('player');

    // Function to fetch and display songs based on the selected genre
    async function fetchSongsByGenre(genre) {
      console.log(`Fetching songs for genre: ${genre}`);
      songContainer.innerHTML = ''; // Clear the previous song list
      const { data, error } = await supabase.storage.from('songs').list(`${genre}/`);
      if (error) {
        console.error('Error fetching songs:', error);
      } else {
        console.log(`Found ${data.length} songs for genre: ${genre}`);
        data.forEach(file => {
          const { publicURL } = supabase.storage.from('songs').getPublicUrl(`${file.path}`);
          console.log('Song public URL:', publicURL);
          const songElement = document.createElement('div');
          const audio = document.createElement('audio');
          audio.src = publicURL;
          audio.controls = true;
          songElement.appendChild(audio);
          songContainer.appendChild(songElement);
        });
      }
    }

    // Event listener for genre selection
    dropdown.addEventListener('change', () => {
      const selectedGenre = dropdown.value;
      fetchSongsByGenre(selectedGenre);
    });

    window.onload = async function() {
      // Get the video URL from the query parameter
      const urlParams = new URLSearchParams(window.location.search);
      const videoUrl = urlParams.get('videoUrl');

      // Check if the video URL is defined and a valid URL
      if (videoUrl && isValidURL(videoUrl)) {
        console.log('Received video URL:', videoUrl);

        // Remove the "uploads/" prefix from the video URL
        const cleanedVideoUrl = videoUrl.replace(/^uploads\//, '');

        if (videoPlayer) {
          videoPlayer.src = decodeURIComponent(cleanedVideoUrl);
          console.log('Set video source:', videoPlayer.src);
          // Create a new instance of Plyr after setting the video source
          const player = new Plyr(videoPlayer);
        }
      } else {
        console.log('No valid video URL received.');
      }

      // Fetch and display songs for the initial genre
      const initialGenre = dropdown.value;
      console.log(`Initial genre: ${initialGenre}`);
      fetchSongsByGenre(initialGenre);
    };

    // Helper function to check if a URL is valid
    function isValidURL(url) {
      try {
        new URL(url);
        return true;
      } catch (_) {
        return false;
      }
    }
  </script>
</body>
</html>

**Comment by GaryAustin1** [comment]
Not going thru all the code, just a couple of things to look at.  

Is the bucket public?  getPublicUrl does not check if the file exists or is accessible it just does a string concatenation.  
You have a list operation on the bucket.  That requires you meet select RLS on storage.objects even if a public bucket.
